pydust/src/pytypes.zig

```diff
@@ -669,12 +669,41 @@ fn RichCompare(comptime definition: type) type {
 
         const compareFuncs = blk: {
             var funcs_: [6]?BinaryFunc = .{ null, null, null, null, null, null };
-            for (funcs.compareFuncs, 0..) |func, i| {
-                if (@hasDecl(definition, func)) {
-                    funcs_[i] = &BinaryOperator(definition, func).call;
+            for (&funcs_, funcs.compareFuncs) |*func, funcName| {
+                if (@hasDecl(definition, funcName)) {
+                    if (std.mem.eql(u8, funcName, "__eq__")) {
+                        func.* = &__eq__;
+                    } else {
+                        func.* = &BinaryOperator(definition, funcName).call;
+                    }
                 }
             }
             break :blk funcs_;
         };
+
+        fn __eq__(pyself: *ffi.PyObject, pyother: *ffi.PyObject) callconv(.C) ?*ffi.PyObject {
+            const typeInfo = @typeInfo(@TypeOf(definition.__eq__));
+            const sig = funcs.parseSignature("__eq__", typeInfo.Fn, &.{});
+
+            if (sig.selfParam == null) @compileError("__eq__ must take a self parameter");
+            if (sig.nargs != 1) @compileError("__eq__ must take exactly one parameter after self parameter");
+
+            const self: *PyTypeStruct(definition) = @ptrCast(pyself);
+            // If other is not PyObject it must be the same type as self.
+            // In that case check that other is of the same type as self to short circuit.
+            if (sig.argsParam != py.PyObject) {
```

Comment:
I think this check should be the other way around. `if (sig.argsParam == *const definition)`

Reply:
It's not a common use-case, but lots of Python "DSLs" override `==` to do funky things. They may want to implement `pub fn __eq__(self: *const Self, other: *const Other) { ... }` and we would incorrectly shortcut/return False in this case.

Reply:
I thought about it and it's a bit of a weird thing to do. If `Other` is a subtype of `Self` then everything works fine. But agreed that this can get weird with subtyping